arangod/Aql/IResearchViewExecutor.tpp

```diff
@@ -394,29 +444,143 @@ void IndexReadBuffer<ValueType, copySorted>::pushValue(
 
 template<typename ValueType, bool copySorted>
 void IndexReadBuffer<ValueType, copySorted>::finalizeHeapSort() {
-  std::sort(
-      _rows.begin(), _rows.end(),
-      BufferHeapSortContext{_numScoreRegisters, _scoresSort, _scoreBuffer});
+  std::sort(_rows.begin(), _rows.end(),
+            BufferHeapSortContext<typename StoredValuesContainer::value_type>{
+                _heapSortValues, _heapSort});
   if (_heapSizeLeft) {
     // heap was not filled up to the limit. So fill buffer here.
     _storedValuesBuffer.resize(_keyBuffer.size() * _storedValuesCount);
   }
 }
 
 template<typename ValueType, bool copySorted>
+template<typename ColumnReaderProvider>
+velocypack::Slice IndexReadBuffer<ValueType, copySorted>::readHeapSortColumn(
+    arangodb::iresearch::HeapSortElement const& cmp, irs::doc_id_t doc,
+    ColumnReaderProvider& readerProvider, size_t readerSlot) {
+  TRI_ASSERT(_heapOnlyStoredValuesReaders.size() >= readerSlot);
+  TRI_ASSERT(!cmp.isScore());
+  if (_heapOnlyStoredValuesReaders.size() == readerSlot) {
+    _heapOnlyStoredValuesReaders.push_back(readerProvider(cmp.source));
+  }
+  auto& reader = _heapOnlyStoredValuesReaders[readerSlot];
+  irs::bytes_view val;
+  TRI_ASSERT(!reader.itr || reader.itr->value() <= doc);
+  if (reader.itr && doc == reader.itr->seek(doc)) {
+    val = reader.value->value;
+  } else {
+    val = ref<irs::byte_type>(VPackSlice::nullSlice());
+  }
+  TRI_ASSERT(_currentDocumentBuffer.capacity() > _currentDocumentBuffer.size());
+  _currentDocumentBuffer.emplace_back(val);
+  if constexpr (copySorted) {
+    return getStoredValue(_currentDocumentBuffer.back(), cmp);
+  } else {
+    _currentDocumentSlices.push_back(
+        getStoredValue(_currentDocumentBuffer.back(), cmp));
+    return _currentDocumentSlices.back();
+  }
+}
+
+template<typename ValueType, bool copySorted>
+template<bool fullHeap, typename ColumnReaderProvider>
+void IndexReadBuffer<ValueType, copySorted>::finalizeHeapSortDocument(
+    size_t idx, irs::doc_id_t doc, std::span<float_t const> scores,
+    ColumnReaderProvider& readerProvider) {
+  auto const heapSortSize = _heapSort.size();
+  size_t heapSortValuesIndex = idx * heapSortSize;
+  size_t j = 0;
+  if constexpr (!copySorted) {
+    TRI_ASSERT(_currentDocumentSlices.size() == _currentDocumentBuffer.size());
+  }
+  if constexpr (fullHeap) {
+    TRI_ASSERT(heapSortValuesIndex < _heapSortValues.size());
+  } else {
+    TRI_ASSERT(heapSortValuesIndex == _heapSortValues.size());
+  }
+  for (auto const& cmp : _heapSort) {
+    if (cmp.isScore()) {
+      if constexpr (fullHeap) {
+        _heapSortValues[heapSortValuesIndex].score = scores[cmp.source];
+      } else {
+        _heapSortValues.push_back(HeapSortValue{.score = scores[cmp.source]});
+      }
+    } else {
```

Comment:
code with deeper scope more difficult to read.
Could we make continue in the first simple case